lib/raxx/session/signed_cookie.ex

```diff
@@ -0,0 +1,206 @@
+      ...> |> SignedCookie.extract(config)
+      {:ok, {:user, 25}}
+
+      iex> config = SignedCookie.config(secret: "eggplant")
+      ...> Raxx.request(:GET, "/")
+      ...> |> Raxx.set_header("cookie", "raxx.session=g2gCZAAEdXNlcmEZ--sfbxaB-IEgUt_NwdmmZpJny9OzOx15D-6uwusW6X1ZY=")
+      ...> |> SignedCookie.extract(config)
+      {:error, :could_not_verify_signature}
+
+  A session can be concluded by marking as expired.
+  For example in response to a users sign out request.
+
+      iex> config = SignedCookie.config(secret: "eggplant")
+      ...> Raxx.response(:no_content)
+      ...> |> SignedCookie.expire(config)
+      ...> |> Raxx.get_header("set-cookie")
+      "raxx.session=; path=/; expires=Thu, 01 Jan 1970 00:00:00 GMT; max-age=0; HttpOnly"
+
+  ## NOTE
+
+  - This module will be extracted from the Raxx project before 1.0 release.
+  - The Rack.Session.Cookie module was the inspiration for this functionality.
+    https://github.com/rack/rack/blob/master/lib/rack/session/cookie.rb
+  """
+
+  @default_cookie_name "raxx.session"
+
+  @enforce_keys [:secret, :cookie_name, :previous_secrets]
+  defstruct @enforce_keys
+
+  @doc """
+  Setup configuration to work with sessions
+
+  ## Options
+
+  - **secret:** (required) a secure random value used for signing session data.
+  - **cookie_name:** default is `"raxx.session"`
+  - **previous_secrets:** acceptable secrets to verify against old sessions.
+  """
+  def config(options) do
+    secret =
+      case Keyword.fetch(options, :secret) do
+        {:ok, secret} ->
+          secret
+
+        :error ->
+          raise "A `:secret` must be set when using signed cookies."
+      end
+
+    cookie_name = Keyword.get(options, :cookie_name, @default_cookie_name)
+    previous_secrets = Keyword.get(options, :previous_secrets, [])
+
+    %__MODULE__{
+      secret: secret,
+      cookie_name: cookie_name,
+      previous_secrets: previous_secrets
+    }
+  end
+
+  @doc """
+  Overwrite a clients session with a new value.
+  """
+  def embed(response, session, session_config = %__MODULE__{}) do
+    payload = safe_encode(session)
+    digest = safe_digest(payload, session_config.secret)
+
+    # Where to put 4096 check
+    # It could be in Raxx.set_cookie
+    response
+    |> Raxx.set_header(
+      "set-cookie",
+      SetCookie.serialize(session_config.cookie_name, payload <> "--" <> digest)
+    )
+  end
+
+  @doc """
+  Extract and verify the session sent from a client.
+  """
+  def extract(request, session_config = %__MODULE__{}) do
+    case Raxx.get_header(request, "cookie") do
+      nil ->
+        {:error, :no_cookies_sent}
+
+      cookie_header ->
+        case Cookie.parse(cookie_header) do
+          cookies = %{} ->
+            case Map.fetch(cookies, "#{session_config.cookie_name}") do
+              {:ok, session_cookie} ->
+                case String.split(session_cookie, "--", parts: 2) do
```

Comment:
This nesting is a bit bleurgh, can this be refactored?

Reply:
Fair comment, though I am going to leave it for the moment. I could come up with some intermediate functions but they are not yet reused anywhere